diff-containers/README.md

```diff
@@ -1 +1,238 @@
+   `Monoid` and `Group`.
+2. Applying diffs to maps should have a nice semantics.
+
+An overview of the sections in this document:
+
+* Section [1) High-level overview](#high-level-overview) provides a high-level
+  overview of diffs, how to create them, and how to apply them.
+* Section [2) A first attempt at implementing
+  diffs](#2-a-first-attempt-at-implementing-diffs) describes a straightforward
+  way to implement diffs, and where this goes wrong if we try to define a
+  `Group` instance.
+* Section [3) Implementing diffs with
+  inverses](#3-implementing-diffs-with-inverses) describes how we implement
+  diffs to facilitate a `Group` instance.
+* Section [4) Applying diffs with inverses: positivity and
+  normality](#4-applying-diffs-with-inverses-positivity-and-normality) describes
+  how applying diffs can fail, but which can be solved by guaranteeing the
+  positivity and normality properties.
+
+### 1) High-level overview
+
+For each of the key-value pairs in a map, the difference type will keep track of
+whether a value is to be inserted or deleted:
+
+```haskell
+newtype Diff k v = Diff {- omitted -}
+
+data DiffEntry v =
+    Insert !v
+  | Delete !v
+  {- omitted for now -}
+```
+
+One method of constructing a diff is to compute the difference between two maps.
+Other methods include creating diffs from maps and lists directly:
+
+```haskell
+diff :: (Ord k, Eq v) => Map k v -> Map k v -> Diff k v
+
+fromMap :: Map k (DiffEntry v) -> Diff k v
+fromList :: Ord k => [(k, DiffEntry v)] -> Diff k v
+```
+
+Diffs can be applied to maps to obtain new maps. For diffs constructed using
+`diff`, the following `property` also holds:
+
+```haskell
+unsafeApplyDiff :: Ord k => Map k v -> Diff k v -> Map k v
+
+property :: Map k v -> Map k v -> Bool
+property = unsafeApplyDiff m1 (diff m1 m2) == m2
+```
+
+### 2) A first attempt at implementing diffs
+
+A straightfoward to represent differences on maps is to keep track of the most
+recent change for each key in the map.
+
+```haskell
+type Diff_ k v = Diff_ (Map k (Delta_ v))
+data Delta_ = Insert_ !v | Delete_ !v
+```
+
+Diffs can then be summed (or combined/`mappend`ed) through a
+`Semigroup`/`Monoid` instance that behaves like a pair-wise `Last` monoid.
+
+```haskell
+   fromList [(1, Insert_ 54), (2, DDelete 42)                ]
+<> fromList [                 (2, Insert_ 17), (3, DDelete 1)]
+== fromList [(1, Insert_ 54), (2, Insert_ 17), (3, DDelete 1)]
+```
+
+What if we want to remove diffs, i.e. through subtraction provided by a `Group`
+instance? In the current definition, there is no way to recover information
+about previous changes, since we only have the last/most recent change. In
+particular, we would like to be able to have the following:
+
+```haskell
+   fromList [(1, Insert_ 54), (2, DDelete 42)                ]
+<> fromList [                 (2, Insert_ 17), (3, DDelete 1)]
+~~ fromList [                 (2, Insert_ 17), (3, DDelete 1)]
```

Comment:
Is `~~` the subtraction op? Maybe we need to clarify this.

Reply:
Also we should probably add some parenthesis or talk about precedence: ie what is true in the code above `(a <> b) ~~` or `a <> (b ~~ c)`?

Reply:
According to the documentation, `(<>)` is right associative and `(~~)` is left associative with a higher precedence. However, since these are monoids (and `(~~)` is an alias for `\x y -> x <> invert y`), the actual parentheses don't matter for the result because of the associativity property

Reply:
I now see that the adding parentheses to the first sum makes the example more clear, so I've updated this